plugins/modules/ec2_instance.py

```diff
@@ -1284,122 +1407,164 @@ def add_or_update_instance_profile(instance, desired_profile_name):
 
-        if params.get("security_group") or params.get("security_groups"):
-            groups = discover_security_groups(
-                group=params.get("security_group"),
-                groups=params.get("security_groups"),
-                subnet_id=spec["SubnetId"],
-            )
-            spec["Groups"] = groups
-        if network.get("description") is not None:
-            spec["Description"] = network["description"]
-        # TODO more special snowflake network things
-
-        return [spec]
-
-    # handle list of `network.interfaces` options
-    for idx, interface_params in enumerate(network.get("interfaces", [])):
-        spec = {
-            "DeviceIndex": idx,
-        }
-
-        if isinstance(interface_params, string_types):
-            # naive case where user gave
-            # network_interfaces: [eni-1234, eni-4567, ....]
-            # put into normal data structure so we don't dupe code
-            interface_params = {"id": interface_params}
-
-        if interface_params.get("id") is not None:
-            # if an ID is provided, we don't want to set any other parameters.
-            spec["NetworkInterfaceId"] = interface_params["id"]
-            interfaces.append(spec)
-            continue
 
-        spec["DeleteOnTermination"] = interface_params.get("delete_on_termination", True)
+def ansible_to_boto3_eni_specification(
+    interface: Dict[str, Any], subnet_id: Optional[str], groups: Optional[Union[List[str], str]]
+) -> Dict[str, Any]:
+    """
+    Converts Ansible network interface specifications into AWS network interface spec
+        Parameters:
+            interface: Ansible network interface specification
+            subnet_id: Subnet Id
+            security_groups: Optional list of security groups.
+            security_group: Optional security group.
+        Returns:
+            AWS network interface specification.
+    """
+    spec = {
+        "DeviceIndex": interface.get("device_index") or 0,
+    }
+    if interface.get("assign_public_ip") is not None:
+        spec["AssociatePublicIpAddress"] = interface["assign_public_ip"]
 
-        if interface_params.get("ipv6_addresses"):
-            spec["Ipv6Addresses"] = [{"Ipv6Address": a} for a in interface_params.get("ipv6_addresses", [])]
+    if interface.get("subnet_id"):
+        spec["SubnetId"] = interface["subnet_id"]
+    elif subnet_id:
+        spec["SubnetId"] = subnet_id
+    else:
+        spec["SubnetId"] = describe_default_subnet(module)
+
+    if interface.get("ipv6_addresses"):
+        spec["Ipv6Addresses"] = [{"Ipv6Address": a} for a in interface.get("ipv6_addresses")]
+
+    if interface.get("private_ip_address"):
+        spec["PrivateIpAddress"] = interface["private_ip_address"]
+
+    if interface.get("private_ip_addresses"):
+        spec["PrivateIpAddresses"] = []
+        for addr in interface.get("private_ip_addresses"):
+            d = {"PrivateIpAddress": addr}
+            if isinstance(addr, dict):
```

Comment:
No, we keep this check because the legacy attribute `network` define `private_ip_address` as a list of `string` while the new `network_interfaces` attribute defines it as a list of `dict`